### PR TITLE
Synchronization: Fences (CPU-GPU) and Semaphores (GPU-GPU)

### PR DESCRIPTION
**Pull request by LeonBrands**

### CODE REVIEW

File: legion/engine/llri-dx/detail/device.cpp [hunk 1]
```diff
@@ -54,4 +54,86 @@ namespace LLRI_NAMESPACE
...
+                    return directx::mapHRESULT(r);
+
+                r = WaitForSingleObject(fence->m_event, static_cast<DWORD>(timeout)); // windows takes the timeout in ms so we can pass it directly
+
+                if (r == WAIT_TIMEOUT)
+                    return result::Timeout;
+
+                if (r == WAIT_FAILED)
+                    return result::ErrorUnknown;
+		    }
```
Comment (GlynLeine) [inline legion/engine/llri-dx/detail/device.cpp:108]: some indenting issue here

File: legion/engine/llri/detail/flags.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace LLRI_NAMESPACE
+{
+    /**
+     * @brief wrapper structure for flags of type E, used to distinguish between x_flag_bits and x_flags.
+    */
+    template<typename E>
+    struct flags
+    {
+        using U = std::underlying_type_t<E>;
```
Comment (GlynLeine) [inline legion/engine/llri/detail/flags.hpp:18]: since U is a public alias, maybe use a more descriptive name?

File: legion/engine/llri/llri.hpp [hunk 3]
```diff
@@ -52,6 +53,9 @@
 #define LLRI_NAMESPACE llri
 #endif
 
+#define LLRI_TIMEOUT_NONE 0u
+#define LLRI_TIMEOUT_INFINITE 0xFFFFFFFFu
```
Comment (GlynLeine) [inline legion/engine/llri/llri.hpp:57]: does this actually end up disabling timeout, or is this just the maximum timeout? cuz if so, LLRI_TIMEOUT_MAX might be a better name

File: legion/engine/llri/detail/flags.hpp (hunk 2 above)
Comment (LeonBrands) [reply legion/engine/llri/detail/flags.hpp:18]: sure! just wasn't sure what to name it

Comment (GlynLeine) [reply legion/engine/llri/detail/flags.hpp:18]: maybe just `underlying_t`

Comment (LeonBrands) [reply legion/engine/llri/detail/flags.hpp:18]: that works

File: legion/engine/llri/llri.hpp (hunk 3 above)
Comment (LeonBrands) [reply legion/engine/llri/llri.hpp:57]: Depends on the implementation. In DirectX it does, Vulkan does not give this same guarantee, but it does guarantee that the time the function takes is finite.

Comment (GlynLeine) [reply legion/engine/llri/llri.hpp:57]: yeah then i'd say call it LLRI_TIMEOUT_MAX, which may be infinite, or might not be

**Review by GlynLeine — APPROVED** [review]

**Pull request merged**